sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala

```diff
@@ -448,6 +448,43 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
       rand(Random.nextLong()), randn(Random.nextLong())
     ).foreach(assertValuesDoNotChangeAfterCoalesceOrUnion(_))
   }
+
+  test("SPARK-21281 fails if functions have no argument") {
+    var errMsg = intercept[AnalysisException] {
+      spark.range(1).select(array())
+    }.getMessage
+    assert(errMsg.contains("due to data type mismatch: input to function coalesce cannot be empty"))
+
+    errMsg = intercept[AnalysisException] {
+      spark.range(1).select(map())
+    }.getMessage
+    assert(errMsg.contains("due to data type mismatch: input to function coalesce cannot be empty"))
+
+    // spark.range(1).select(coalesce())
+    errMsg = intercept[AnalysisException] {
+      spark.range(1).select(coalesce())
+    }.getMessage
+    assert(errMsg.contains("due to data type mismatch: input to function coalesce cannot be empty"))
+
+    // This hits java.lang.AssertionError
+    // spark.range(1).select(struct())
+
+    errMsg = intercept[IllegalArgumentException] {
+      spark.range(1).select(greatest())
+    }.getMessage
+    assert(errMsg.contains("requirement failed: greatest requires at least 2 arguments"))
```

Comment:
Nit: Exception types are different between `sql.functions.greatest` and `selectExpr("greatest()")`;
```
cala> spark.range(1).select(greatest()).show
java.lang.IllegalArgumentException: requirement failed: greatest requires at least 2 arguments.
  at scala.Predef$.require(Predef.scala:224)
  at org.apache.spark.sql.functions$.greatest(functions.scala:1570)
  ... 48 elided

scala> spark.range(1).selectExpr("greatest()").show
org.apache.spark.sql.AnalysisException: cannot resolve 'greatest()' due to data type mismatch: GREATEST requires at least 2 arguments; line 1 pos 0;
'Project [unresolvedalias(greatest(), Some(<function1>))]
+- Range (0, 1, step=1, splits=Some(4))

  at org.apache.spark.sql.catalyst.analysis.package$AnalysisErrorAt.failAnalysis(package.scala:42)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$checkAnalysis$1$$anonfun$apply$2.applyOrElse(CheckAnalysis.scala:95)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$checkAnalysis$1$$anonfun$apply$2.applyOrElse(CheckAnalysis.scala:87)
  at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformUp$1.apply(TreeNode.scala:289)
  at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$transformUp$1.apply(TreeNode.scala:289)
  at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:70)
```